### ISSUE TITLE
Is the agent vector necessary in implementation?

### ISSUE THREAD

**Issue by siberiamark** [happynear/NormFace#6]
How about take the softmax center vector as the agent vector, i.e.  sharing the weight of "id_weight_ip" and "id_weight_l2".? Intuitively I think it makes sense , and the removal of agent vector doesn't harm the concept of metric learning, it's just that the mutual influence of the two losses is modeled in an implicit way.What do you think ....

**Comment by happynear** [comment]
Please notice that after normalization, 2 - 2 * D_ip = D_l2 is only a linear transform. If softmax loss and C-contrastive loss share the same distance layer, this is actually another **single** loss function. If I use different distance layer for the two loss functions, there will be two kinds of agent. The agent of softmax loss will have different behavior compared with the agent of C-contrastive loss. I think this implementation makes more sense.

Anyway, the accuracy gain of joint training strategy is so small that you may not pay too much attention to this. The key is the normalization. If you are more interested in the algorithm than the performance, please let me know. I can talk more with you about the algorithm.

**Comment by siberiamark** [comment]
OK I get your point.  Just as you mentioned in the paper, when m = 0 , C-agent converge to softmax agent. What makes the difference is the parameter m, to my understanding, as long as m>0,sharing the agent doesn't cancel the influence of contrastive loss.

**Comment by siberiamark** [comment]
Though the key point is to performance is normalize the weight and feature, I think solving the problem of sampling pairs and triplets is of more contribution. The paper is really well written, hope it will be accepted.

**Comment by happynear** [comment]
When m=0, C-contrastive loss is equavalent to center loss, not softmax loss. Softmax's agents play a different, but similar role with the agents of C-contrastive loss.

**Comment by jay2002** [comment]
@happynear In your paper, features and weights are normalized before c-contrastive loss. Have you try to apply c-constrasrive loss on the original features and original weights?

**Comment by happynear** [comment]
@jay2002, I only tried it on mnist. It will be more difficult to set the margin parameter if not normalized. You need to suspend training every several epoches and set the new margin. I don't like such kind of training scheme.

**Comment by jay2002** [comment]
@happynear thanks.
Actually I have tried center loss with and w/o l2-norm, and I found that if we set the hyper-params carefully, w/o l2-norm performs better on center loss. That's why I'm curious about c-constrasrive w/o l2-norm...

**Comment by happynear** [comment]
"on center loss" or "on ResNet"?

**Comment by jay2002** [comment]
Sorry for my poor english ... In my experiments.
fully-connected feature + center loss > l2 normed fully-connected feature + center loss
为了不引起歧义我再用中文说一遍好了……
我的实验现象就是，fc特征加yandong原版的center loss（A）会比fc特征做l2-norm再接原版的center loss（B）效果好不少。所以我比较好奇fc和center都做l2-nrom（C），以及fc和center都不归一化的c-contrastive（D），和A、B相比，效果会怎么样。所以就想问问你你有没做过这样的实验

**Comment by happynear** [comment]
I just use center loss to create baseline model. I didn't try normalization on center loss.
